### PR TITLE
Update Kotlin + kotlinx-metadata to 1.5 and 0.3.0

### PR DESCRIPTION
**Pull request by ZacSweers**
Updates a few other libraries along the way and updates updates APIs accordingly (namely around JvmDefault and inline class changes).

Also moves all `TypeName` subtypes to separate top-level files since this restriction on sealed types is removed in Kotlin 1.5.

Commits are intended to be individually reviewable!

### CODE REVIEW

**Review by Egorand — APPROVED** [review]
Should we cut a release?

File: build.gradle.kts [hunk 1]
```diff
@@ -39,7 +39,6 @@ subprojects {
   }
   tasks.withType<KotlinCompile> {
     kotlinOptions {
-      jvmTarget = "1.8"
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply build.gradle.kts:42]: Yup

**Comment by ZacSweers** [comment]
I'm gonna take another spin at #1054 if you want to hold off a bit, fine with releasing now though too.

**Comment by Egorand** [comment]
No rush, can definitely get some more fixes into the release

**Pull request merged**